### PR TITLE
Add editor route to fix browser navigation to/from MakeCode

### PR DESCRIPTION
**Pull request by microbit-robert**

### CODE REVIEW

File: src/pages/EditorPage.tsx [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    } else {
+      setEditorOpen(true);
+    }
+
+    return () => {
+      if (model) {
+        setEditorOpen(false);
+      }
+    };
+  }, [model, navigate, setEditorOpen]);
```
Comment (microbit-robert) [inline src/pages/EditorPage.tsx:23]: Not a fan of this but it will catch using the browser forward/back buttons, and navigation by URL / link.

Comment (microbit-robert) [reply src/pages/EditorPage.tsx]: As discussed `setEditorOpen(false)` now handles the model condition internally.

File: src/hooks/project-hooks.tsx [hunk 2]
```diff
@@ -79,27 +84,54 @@ export const ProjectProvider = ({
...
+      if (!doAfterEditorUpdatePromise.current && checkIfProjectNeedsFlush()) {
+        doAfterEditorUpdatePromise.current = new Promise<void>(
+          (resolve, reject) => {
+            // driverRef.current is not defined on first render.
+            // Only an issue when navigating to code page directly.
+            if (!driverRef.current) {
+              reject(
+                new TypeError(
+                  "Cannot read properties of null (reading 'importProject')"
+                )
```
Comment (microbit-matt-hillsdon) [inline src/hooks/project-hooks.tsx:105]: This feels like it might be very confusing to future-us. Can we reject with something that's clearly custom?

Comment (microbit-matt-hillsdon) [reply src/hooks/project-hooks.tsx:105]: If we threw a custom error subclass we could check for it below.

Comment (microbit-matt-hillsdon) [reply src/hooks/project-hooks.tsx:105]: Otherwise happy with this.

File: src/hooks/project-hooks.tsx [hunk 3]
```diff
@@ -108,14 +138,24 @@ export const ProjectProvider = ({
...
+  const browserNavigationToEditor = useCallback(async () => {
+    try {
+      await doAfterEditorUpdate(() => {
+        return Promise.resolve();
+      });
+      return true;
+    } catch (e) {
+      // doAfterEditorUpdate only fails when navigating to the code page directly.
+      // In this case, the caller of browserNavigationToEditor redirects.
+      return false;
```
Comment (microbit-robert) [inline src/hooks/project-hooks.tsx:154]: We could now check for the right CodeEditorError here, but it's less clear what to do if it's an unexpected error. Navigate to a different page anyway, or possibly reload the page.

File: src/hooks/project-hooks.tsx (hunk 2 above)
Comment (microbit-robert) [reply src/hooks/project-hooks.tsx]: Added a `CodeEditorError` subclass with clearer message. Haven't checked for it though. See comment below. [5d0fab3](https://github.com/microbit-foundation/ml-trainer/pull/416/commits/5d0fab3ef2cea8eeea3799ea0b9d7fae2780c344)

File: src/hooks/project-hooks.tsx (hunk 3 above)
Comment (microbit-matt-hillsdon) [inline src/hooks/project-hooks.tsx:154]: Maybe roughly this? I think most of the value is in having CodeEditorError in the two places to make the flow clearer without relying on comments.

```
if (e instanceof CodeEditorError) {
  // Comment as above
  return false;
}
// Unexpected error, can't handle better than the redirect
logging.error(e)
return false;
```

Comment (microbit-robert) [reply src/hooks/project-hooks.tsx:159]: [a8ab3da](https://github.com/microbit-foundation/ml-trainer/pull/416/commits/a8ab3dad585aa59ebea6831537eacd17dd76a9db)

**Review by microbit-matt-hillsdon — APPROVED** [review]
LGTM

**Pull request merged**